File: 4 kyu/Human readable duration format/solution.rb
def format_duration(total)
  if total == 0
    "now"
  else
    duration = {
      year:   total / (60 * 60 * 24 * 365),
      day:    total / (60 * 60 * 24) % 365,
      hour:   total / (60 * 60) % 24,
      minute: total / 60 % 60,
      second: total % 60
    }
  
    @output = []
  
    duration.each do |key, value|
      if value > 0
        @output << "#{value} #{key}"
        @output.last << "s" if value != 1
      end
    end
  
    @output.join(', ').gsub(/\,\s(?=\d+\s\w+$)/, " and ")
  end
end

__________________________________________________
def format_duration(seconds)
  return "now" if seconds == 0
  m,s = seconds.divmod(60)
  h,m = m.divmod(60)
  d,h = h.divmod(24)
  y,d = d.divmod(365)

  *f,l = {'year'=>y,'day'=>d,'hour'=>h,'minute'=>m,'second'=>s}.to_a.select{|i| i.last > 0}.map{|i| (i.last>1)? "#{i.last} #{i.first+'s'}": "#{i.last} #{i.first}"}
  (f.count > 0)? f.join(', ') + ' and '+l : l
end

__________________________________________________
def format_duration(seconds)
  return "now" if seconds.zero?

  durations = {
    years: 31_536_000,
    days:      86_400,
    hours:      3_600,
    minutes:       60,
    seconds:        1
  }

  singularize = ->((k, v)) { "#{v} #{v == 1 ? k[0..-2] : k}" }

  units = durations.each_with_object({}) do |(name, duration), units|
    units[name], seconds = seconds.divmod(duration)
  end.reject { |k, v| v.zero? }.map(&singularize)

  last = units.pop
  return last if units.empty?
  units.join(", ") << " and #{last}"
end

File: 4 kyu/Human readable duration format/solution.clj
52742f58faf5485cae000b9a


(ns human-readable
  (:require clojure.string))

(defn- pluralize [[unit v]]
  (when (and v (pos? v))
    (str v " " unit (when (> v 1) "s"))))

(defn formatDuration [secs]
  (if (zero? secs)
    "now"
    ; else
    (-> {"second" 60   ; seconds-in-minute
         "minute" 60   ; minutes-in-hour
         "hour"   24   ; hours-in-day
         "day"    365  ; days-in-year
         "year"   nil} ; years-in-eternity
      (->> 
        (reduce
          (fn [[C qty]
               [unit d]]
            (let [modulus (if d (mod qty d) qty)]
              [(cond-> C
                 (pos? modulus)
                 (assoc unit modulus))
               (when d
                 (quot qty d))]))
          [{} secs])
        first
        (map pluralize)
        reverse
        (clojure.string/join ", "))
      (clojure.string/replace #", ([^,]*)$" " and $1"))))

___________________________________________________
(ns human-readable)

(def units
  [{:name "year" :unit-duration 31536000}
   {:name "day" :unit-duration 86400}
   {:name "hour" :unit-duration 3600}
   {:name "minute" :unit-duration 60}
   {:name "second" :unit-duration 1}])

(defn to-units [secs]
  (loop [remaining secs
         [unit & more-units] units
         results []]
    (if unit
      (let [{name :name unit-duration :unit-duration} unit
            unit-count (quot remaining unit-duration)]
        (recur
          (mod remaining unit-duration)
          more-units
          (if (pos? unit-count)
            (conj results (str unit-count " " name (when (< 1 unit-count) "s")))
            results)))
      results)))

(defn formatDuration [secs]
  (if-not (pos? secs)
    "now"
    (let [durations (to-units secs)]
      (str (when (< 1 (count durations)) (str (clojure.string/join ", " (butlast durations)) " and ")) (last durations)))))

___________________________________________________
(ns human-readable)

(use '[clojure.string :only (join blank?)])

(defn formatDuration [secs]
  (let [units [[60 "second"] [60 "minute"] [24 "hour"] [365 "day"] [(inc secs) "year"]]
        extract-time (fn [[_ s _] [unit name]]
                       (let [time (rem s unit)
                             rem-time (quot s unit)
                             name (join [name (if (== 1 time) "" "s")])]
                         [time rem-time name]))
        time-parts (->> (reductions extract-time [0 secs ""] units)
                        (filter #(pos? (first %)))
                        (map (fn [[s _ n]] (join [s " " n])))
                        reverse)]
    (if (zero? (count time-parts))
      "now"
      (join " and " (filter #(not (blank? %)) [(join ", " (butlast time-parts)) (last time-parts)])))
    ))
